### PR TITLE
Adds new authentication method for model management API

### PR DESCRIPTION
**Pull request by dustins**
*Issue #, if available:*

*Description of changes:*
Starting with the 3.0 release, which introduced dynamic model management, we required admin authentication through an IdP to manage models. However, this approach doesn't work in certain scenarios. To address this, the PR introduces a new authentication method using the model management token secret.

*Testing:*
- I am able to create a model using the Model Management UI 
- I am able to create a model by submitting a POST request to the APIGW endpoint `/model` using the model management token
- I am not able create a model by submitting a POST request to the APIGW endpoint `/model` without the management token

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: test/cdk/stacks/core-api-base.test.ts [hunk 1]
```diff
@@ -112,6 +112,6 @@ describe.each(regions)('API Core Nag Pack Tests | Region Test: %s', (awsRegion)
 
     test('NIST800.53r5 CDK NAG Errors', () => {
         const errors = Annotations.fromStack(stack).findError('*', Match.stringLikeRegexp('NIST.*'));
-        expect(errors.length).toBe(6);
+        expect(errors.length).toBe(7);
```
Comment (dustins) [inline test/cdk/stacks/core-api-base.test.ts:115]: This change is a complaint that the authorizer is using an inline policy as a result of adding in `managementKeySecret.grantRead(authorizerLambda);`

File: lambda/authorizer/lambda_functions.py [hunk 2]
```diff
@@ -134,3 +143,30 @@ def is_admin(jwt_data: dict[str, Any], admin_group: str, jwt_groups_property: st
         else:
             return False
     return admin_group in current_node
+
+
+def refresh_management_tokens() -> list[str]:
+    """Return secret management tokens if they exist."""
+    global secret_tokens
```
Comment (petermuller) [inline lambda/authorizer/lambda_functions.py:150]: lets rename to `get_management_tokens` and then remove the global variable since we don't have a "global" use case based on the code here

File: lambda/authorizer/lambda_functions.py [hunk 3]
```diff
@@ -18,13 +18,17 @@
...
 
+import boto3
 import create_env_variables  # noqa: F401
 import jwt
 import requests
 from utilities.common_functions import authorization_wrapper, get_id_token
 
 logger = logging.getLogger(__name__)
 
+secrets_manager = boto3.client("secretsmanager", region_name=os.environ["AWS_REGION"])
```
Comment (petermuller) [inline lambda/authorizer/lambda_functions.py:29]: let's add the retry config in here too- we have a method for that, we just need to import it. examples:
https://github.com/awslabs/LISA/blob/develop/lambda/models/lambda_functions.py#L24
https://github.com/awslabs/LISA/blob/develop/lambda/models/lambda_functions.py#L64-L67

File: lambda/authorizer/lambda_functions.py [hunk 4]
```diff
@@ -134,3 +143,30 @@ def is_admin(jwt_data: dict[str, Any], admin_group: str, jwt_groups_property: st
...
+                    SecretId=os.environ.get("MANAGEMENT_KEY_NAME"), VersionStage="AWSCURRENT"
+                )["SecretString"]
+            )
+            secret_tokens.append(
+                secrets_manager.get_secret_value(
+                    SecretId=os.environ.get("MANAGEMENT_KEY_NAME"), VersionStage="AWSPREVIOUS"
+                )["SecretString"]
+            )
+        except Exception:
+            logger.info(f"No previous secret version for {os.environ.get('MANAGEMENT_KEY_NAME')}")
```
Comment (petermuller) [inline lambda/authorizer/lambda_functions.py:165]: can we add the stack trace or actual error in the output too? if we get an exception for something that is not related to getting the previous key then we're outputting a possibly misleading error message.

File: lambda/authorizer/lambda_functions.py [hunk 5]
```diff
@@ -134,3 +143,30 @@ def is_admin(jwt_data: dict[str, Any], admin_group: str, jwt_groups_property: st
...
+
+    return secret_tokens
+
+
+def is_valid_management_token(id_token: str) -> bool:
```
Comment (petermuller) [inline lambda/authorizer/lambda_functions.py:170]: from one of my other comments, we may be able to remove this

File: lib/api-base/authorizer.ts [hunk 6]
```diff
@@ -82,12 +86,16 @@ export class CustomAuthorizer extends Construct {
...
                 JWT_GROUPS_PROP: config.authConfig!.jwtGroupsProperty,
+                MANAGEMENT_KEY_NAME: managementKeySecretNameStringParameter.stringValue
             },
             role: role,
             vpc: vpc,
             securityGroups: securityGroups,
         });
 
+        const managementKeySecret = Secret.fromSecretNameV2(this, createCdkId([id, 'managementKey']), managementKeySecretNameStringParameter.stringValue);
+        managementKeySecret.grantRead(authorizerLambda);
```
Comment (petermuller) [inline lib/api-base/authorizer.ts:97]: is this correct? or should the read be granted to the authorizer lambda role? i normally would have expected that the execution role gets the permissions

File: lib/serve/index.ts [hunk 7]
```diff
@@ -46,6 +46,8 @@ export class LisaServeApplicationStack extends Stack {
     public readonly restApi: FastApiContainer;
     public readonly modelsPs: StringParameter;
     public readonly endpointUrl: StringParameter;
+    public readonly managementKeySecret: Secret;
+    public readonly managementKeySecretNameStringParameter: StringParameter;
```
Comment (petermuller) [inline lib/serve/index.ts:50]: why do we need to add these as externally readable members? i'd REALLY like to avoid exposing these, and instead just contrive the values in code, like you already did. I don't see any external references to these values (which is good), but because of the dangers of adding cross-stack dependencies that can block cdk updates (especially with constantly rotating values), I would much prefer to not expose these at all and keep them internal to this stack. (please see this change as a reference for my aversion to cross-stack param references: https://github.com/awslabs/LISA/commit/e2ae9efa5e169354f24f7d81ffac8007c6599d52 )

**Review by petermuller — CHANGES_REQUESTED** [review]

File: lambda/utilities/common_functions.py [hunk 8]
```diff
@@ -285,8 +285,12 @@ def get_id_token(event: dict) -> str:
...
         raise ValueError("Missing authorization token.")
 
-    token = auth_header[1]
-    return str(token)
+    if len(auth_header) == 1:
+        # secret management token won't be split into multiple segments
+        return str(auth_header[0])
+    else:
+        # Bearer tokens will have the token in the second segment
+        return str(auth_header[1])
```
[the remark this answers is not in the document]
Comment (dustins) [reply lambda/utilities/common_functions.py:293]: I agree that is much better.

File: lib/api-base/authorizer.ts (hunk 6 above)
Comment (dustins) [reply lib/api-base/authorizer.ts:97]: It does add it to the lambda role. I can try adding it for the role directly though.

Comment (dustins) [reply lib/api-base/authorizer.ts:97]: Actually since `role` is nullable and it works like this it is probably best to keep it as is since we know that the `authorizerLambda` will never be null.

File: lambda/authorizer/lambda_functions.py [hunk 9]
```diff
@@ -134,3 +144,24 @@ def is_admin(jwt_data: dict[str, Any], admin_group: str, jwt_groups_property: st
...
     return admin_group in current_node
+
+
+@cache
+def get_management_tokens() -> list[str]:
+    """Return secret management tokens if they exist."""
+    secret_tokens: list[str] = []
+    secret_id = os.environ.get("MANAGEMENT_KEY_NAME")
+
+    if len(secret_tokens) == 0:
```
Comment (petermuller) [inline lambda/authorizer/lambda_functions.py:155]: this is guaranteed to be 0, so we can remove this check

File: lambda/utilities/common_functions.py [hunk 10]
```diff
@@ -279,14 +280,14 @@ def get_id_token(event: dict) -> str:
...
     elif "Authorization" in event["headers"]:
-        auth_header = event["headers"]["Authorization"].split(" ")
+        auth_header = event["headers"]["Authorization"]
     else:
         raise ValueError("Missing authorization token.")
 
-    token = auth_header[1]
-    return str(token)
+    # remove bearer token prefix if present
+    return re.sub(r"Bearer ", "", str(auth_header), flags=re.I).strip()
```
Comment (petermuller) [inline lambda/utilities/common_functions.py:290]: I'd prefer the removeprefix string method since it's a bit easier to read than this and doesn't import a library, but not a blocking comment. this still does the job- it's a stretch, but this may be a problem if somehow a user has a token that also contains "Bearer " in it (largely unlikely, but possible)

```
>>> import re
>>> re.sub(r"Bearer ", "", "Bearer Bearer 1", flags=re.I)
'1'
```

vs

```
>>> "Bearer Bearer 1".removeprefix("Bearer ")
'Bearer 1'
```

something to consider, i'll accept either way though- i'm sure secrets manager wouldn't give us that string specifically, and spaces probably shouldn't make it into an auth header definition

File: lib/api-base/authorizer.ts (hunk 6 above)
Comment (petermuller) [reply lib/api-base/authorizer.ts:97]: TIL. nice!

File: lambda/utilities/common_functions.py (hunk 10 above)
Comment (dustins) [reply lambda/utilities/common_functions.py:290]: I'll push the change with `removeprefix` but since there is no capitalization guarantees so it is actually between the two below.

```
>>> import re
>>> re.sub(r"^Bearer ", "", id_token, flags=re.I)
``` 

and

```
>>> id_token.removeprefix("Bearer ").removeprefix("bearer ")
```

File: lambda/authorizer/lambda_functions.py [hunk 11]
```diff
@@ -16,15 +16,20 @@
...
 import create_env_variables  # noqa: F401
 import jwt
 import requests
-from utilities.common_functions import authorization_wrapper, get_id_token
+from botocore.exceptions import ClientError
+from utilities.common_functions import authorization_wrapper, get_id_token, retry_config
 
 logger = logging.getLogger(__name__)
 
+secrets_manager = boto3.client("secretsmanager", config=retry_config)
```
Comment (petermuller) [inline lambda/authorizer/lambda_functions.py:31]: region should still be here though, right?

Comment (petermuller) [reply lambda/authorizer/lambda_functions.py]: added it back in next commit.

**Review by petermuller — APPROVED** [review]
